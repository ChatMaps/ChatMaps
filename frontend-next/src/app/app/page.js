"use client"
import { useState, useEffect } from 'react'
import {Map, Marker, ZoomControl} from "pigeon-maps"
import { Form, useForm } from "react-hook-form";
import { app } from "../api/firebase-config";
import { getDatabase, ref, onValue, get, set, remove} from "firebase/database";
import { useBeforeunload } from 'react-beforeunload';
var database = getDatabase(app)


// Data Types

// Chat Message
function Chat({chatObj}) {
  let dateOptions = {
    weekday: 'long',
    year: 'numeric',
    month: 'short',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit'
  };
  return (
    <div className='width-[100%] bg-white rounded-lg mt-1 text-left p-1 grid grid-cols-2 mr-2'>
      <div>
        {chatObj.user}: {chatObj.body}
      </div>
      <div className='text-right text-[#d1d1d1]'>
        {new Date(chatObj.timestamp).toLocaleString(dateOptions)}
      </div>
    </div>
  )
}

// System Chat Message
function SystemMessage({chatObj}) {
  let dateOptions = {
    weekday: 'long',
    year: 'numeric',
    month: 'short',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit'
  };
  return (
    <div className='width-[100%] bg-white rounded-lg mt-1 text-left p-1 grid grid-cols-2 mr-2'>
      <div className='text-[#d1d1d1]'>
        {chatObj.user} has {chatObj.body} the room.
      </div>
      <div className='text-right text-[#d1d1d1]'>
        {new Date(chatObj.timestamp).toLocaleString(dateOptions)}
      </div>
    </div>
  )
}

// Member for Active/Room members in sidebar
function Member({memberObj}) {
  return (
    <div className='cursor-pointer g-[aliceblue] rounded-lg m-3 shadow-xl p-2'>
      {memberObj.username}
    </div>
  )
}

// Chat Room for myRooms and Nearby in sidebar
function ChatRoomSidebar({roomObj, click}) {
  // TODO: Gross fix but it works
function clicker() {
  click(roomObj.name+"-"+roomObj.timestamp, roomObj)
}
return (
  <div onClick={clicker} className='border-[black] border-1 shadow-lg p-2 m-2 rounded-lg cursor-pointer'>
    <div className='col-span-2'>
      <div className='font-bold'>{roomObj.name}</div>
      <div className='italic'>{roomObj.description}</div>
    </div>
  </div>
)
}

// Map module for main page and chat room sidebar
// TODO: MAKE NOT MOVABLE
function Geo({loc, zoom, movable, marker}) {
  if (loc) {
    return (
      <Map center={[loc.latitude, loc.longitude]} defaultZoom={zoom}>
        {marker && <Marker width={50} anchor={[loc.latitude, loc.longitude]} color="red"/>}
        {zoom && <ZoomControl />}
      </Map>
    )
  } else {
    return (
      <Map className="rounded-lg" defaultCenter={[0, 0]} defaultZoom={zoom}/>
    )
  }
  
}

// Module for Welcome Message on main tab landing page
function WelcomeMessage() {
  const [data, setData] = useState(null)
  const [isLoading, setLoading] = useState(true)
  useEffect(() => {
    fetch('/api/user')
      .then((res) => res.json())
      .then((data) => {
        setData(data)
        setLoading(false)
      })
  }, [])
  if (isLoading) return <div></div>
  if (!data) return <div></div>

  return (
    <div className="bg-white rounded-lg m-2 mt-4 text-left p-2 pl-5">
      <div>
        Welcome, {data.firstName} {data.lastName} ({data.username})
      </div>
      <div>
        Lets see what&apos;s happening in your area.
      </div>
    </div>
  )

}

// Main Tabs
// Primary App Landing Page
function MainTabHome({loc}) {
  return (
    <>
    <WelcomeMessage />
    <div className='h-[calc(100%-110px)] m-5 rounded-lg'>
      <Geo loc={loc} zoom={14} movable={true} marker={true}/>
    </div>
    </>
  )
}

// Chatroom Module for Primary Tab
function MainTabChatRoom({roomObj}) {
  var { register, control, reset, handleSubmit} = useForm()
  const [chats, setData] = useState(null)
  const [isLoading, setLoading] = useState(true)

  // Message updater
  useEffect(() => {
    onValue(ref(database, `/rooms/${roomObj.path+"/"+roomObj.name+"-"+roomObj.timestamp}/chats`), (snapshot) => {
        var chatsArr = []
        var messages = snapshot.val()
        for (var message in messages) {
          if (messages[message].isSystem) {
            chatsArr.push(<SystemMessage chatObj={messages[message]} key={messages[message].timestamp}/>)
          } else {
            chatsArr.push(<Chat chatObj={messages[message]} key={messages[message].timestamp}/>)
          }
        }
        setData(chatsArr.reverse())
        setLoading(false)
    })
  }, [])

  function sendMessage(data) {
    reset()
    fetch('/api/user').then((res) => res.json())
    .then((user) => {
      var payload = {
        body: data.message,
        user: user.username,
        isSystem: false,
        timestamp: new Date().getTime()
      }
      set(ref(database,`/rooms/${roomObj.path+"/"+roomObj.name+"-"+roomObj.timestamp}/chats/${new Date().getTime()}-${user.username}`), payload)
    })
  }

  if (isLoading) return <div>Loading</div>
  if (!chats) return <div>No Chats</div>
  return (
    <div className='m-1 h-[100%] rounded-lg'>
      <div className='h-[90%] m-4 overflow-y-auto flex flex-col-reverse'>
        {chats}
      </div>
      <div className='m-2 h-[10%] w-[100%] bg-white rounded-lg'>
        <Form onSubmit={handleSubmit(sendMessage)} control={control} className='w-[100%] p-[0px]'>
          <input type="text" {...register("message")} placeholder="Enter message" className='w-[83%] border-[0px] mt-[8px] mb-[8px]'/>
          <button className="p-2 cursor-pointer bg-[#dee0e0] bg-cyan-500 text-white font-bold rounded-full mr-5 w-[8%]">Send</button>
        </Form>
      </div>
    </div>
  )
}

// Contains most everything for the app homepage
// 
function Home() {
  // It's time to document and change these awful variable names
  // State variables for app page
  const [mainTab, setMainTab] = useState("home") // Primary tab
  const [tab, setTab] = useState("nearby") // Sidebar Tab
  const [chatRoomObj, setChatRoomObj] = useState(null) // Current chatroom object
  const [myRoomsObj, setMyRoomsObj] = useState(null) // My Rooms Object
  const [myRooms, setRoomData] = useState(null) // Current user saved rooms list
  const [isRoomLoading, setRoomLoading] = useState(true) // myRooms loading variable, true = myRooms loading, false = finished loading
  const [isMyRoom, setIsMyRoom] = useState(false) // Is current room in myRooms? true, false
  const [location, setLocation] = useState(null) // location variable [lat,long]
  const [loadingLoc, setLoadingLoc] = useState(true) // location variable loading, true = loading, false = finished loading
  const [nearby, setNearby] = useState(null); // nearby rooms array
  const [loadingNearby, setLoadingNearby] = useState(true); // loading nearby rooms array, true = loading, false = finished loading
  const [chatroomOnline, setChatRoomOnline] = useState(null) // holds online users
  const [chatroomUsers, setChatroomUsers] = useState(null) // holds all chatroom users
  const [users, setUsers] = useState(null) // all users from firebase
  const [alreadyLeft, setAlreadyLeft] = useState(false) // if already left from room

  // Grabs user data, saves to user, then lists the users saved rooms
  useEffect(() => {
    fetch('/api/user').then((res) => res.json())
    .then((user) => {
      get(ref(database, '/users/'+user.uid+'/rooms')).then((snapshot) => {
        var rooms = snapshot.val()
        setMyRoomsObj(rooms)
        var roomArr = []
        for (var room in rooms) {
          var newRoom = <ChatRoomSidebar roomObj={rooms[room]} key={rooms[room].timestamp} click={selectChatRoom}/>
          roomArr.push(newRoom)
        }
        setRoomData(roomArr)
        setRoomLoading(false)
      })
    })
}, [])

  // Grabs the user location
  useEffect(() => {
    if('geolocation' in navigator) {
      // Retrieve latitude & longitude coordinates from `navigator.geolocation` Web API
      navigator.geolocation.getCurrentPosition(({ coords }) => {
        setLocation(coords)
        setLoadingLoc(false)
        var nearbyArr = []
        var path = String(coords.latitude.toFixed(2)).replace(".","")+"/"+String(coords.longitude.toFixed(2)).replace(".","")
        get(ref(database, `/rooms/${path}`)).then((snapshot) => {
          if (snapshot.exists()) {
            var data = snapshot.val()
            for (var room in data) {
              nearbyArr.push(<ChatRoomSidebar roomObj={data[room]} click={selectChatRoom}/>)
            }
            setLoadingNearby(false)
            setNearby(nearbyArr)
          } else {
            setLoadingNearby(false)
          }
        })
      })
    }
  }, []);

  // Grab list of all users
  useEffect(() => {
    get(ref(database, `/users`)).then((snapshot) => {
      setUsers(snapshot.val())
    })
    
  }, []);

  useEffect(() => {
   if (myRoomsObj && chatRoomObj) {
    var roomName = chatRoomObj.name+"-"+chatRoomObj.timestamp
    if (myRooms != null && roomName in myRoomsObj) {
      // its in there
      setIsMyRoom(true)
    } else {
      // its not in there
      setIsMyRoom(false)
    }
   }
  }, [chatRoomObj])

  // CreateRoom Module for Sidebar Create Tab
  function CreateRoom({loc}) {
    var { register, control, reset, handleSubmit} = useForm()
  
    function createRoom(data) {
      reset()
      var path = String(loc.latitude.toFixed(2)).replace(".","")+"/"+String(loc.longitude.toFixed(2)).replace(".","")
      var timestamp = new Date().getTime()
      var payload = {
        name: data.name,
        description: data.description,
        timestamp: timestamp,
        latitude: loc.latitude,
        longitude: loc.longitude,
        path: path
      }
      set(ref(database,`/rooms/${path}/${data.name}-${timestamp}`), payload)
    }
  
    return (
      <div className='overflow-y-auto h-[90%]'>
        <Form control={control} onSubmit={handleSubmit(createRoom)}>
          <input {...register("name")} placeholder='Room Name' className='mt-2'/>
          <input {...register("description")} placeholder='Room Description' className='mt-2'/><br/>
          <div className='mt-3 mb-2'>
            Creating room near ({loc.latitude.toFixed(2)}, {loc.longitude.toFixed(2)})
          </div>
          <button className="p-2 cursor-pointer bg-[#dee0e0] bg-cyan-500 text-white font-bold rounded-full mr-5">Create</button>
        </Form>
      </div>
    )
  }

  // Selects chat room
  function selectChatRoom(roomName, roomObj) {
    fetch('/api/user').then((res) => res.json())
    .then((user) => {
      // Path of chatroom
      var path = roomObj.path+"/"+roomObj.name+"-"+roomObj.timestamp

      setChatRoomObj(roomObj)

      // Send entered message
      var payload = {
        body: "entered",
        user: user.username,
        isSystem: true,
        timestamp: new Date().getTime()
      }
      set(ref(database,`/rooms/${path}/chats/${new Date().getTime()}-${user.username}`), payload)
      
      // Code for Room Data
      set(ref(database, `/rooms/${path}/users/online/${user.uid}`), user)
      onValue(ref(database, `/rooms/${path}`), (snapshot) => {

        // Active users list
        if (snapshot.val().hasOwnProperty("users") && snapshot.val().users.hasOwnProperty("online")) {
          var activeUsers = []
          var activeUsersJSON = snapshot.val().users.online
          for (var user in activeUsersJSON)
            activeUsers.push(<Member memberObj={activeUsersJSON[user]}/>)
          setChatRoomOnline(activeUsers)
        }

        // Users who added to "my rooms"
        if (snapshot.val().hasOwnProperty("users") && snapshot.val().users.hasOwnProperty("all")) {
          var allUsers = []
          var allUsersJSON = snapshot.val().users.all
          for (var user in allUsersJSON)
            allUsers.push(<Member memberObj={allUsersJSON[user]}/>)
          setChatroomUsers(allUsers)
        }

      })
      setMainTab("chat")
      setAlreadyLeft(false)
    })
  }

  // Closes chat room
  function closeChatRoom(roomObj) {
    fetch('/api/user').then((res) => res.json())
    .then((user) => {
      var path = roomObj.path+"/"+roomObj.name+"-"+roomObj.timestamp
      var payload = {
        body: "left",
        user: user.username,
        isSystem: true,
        timestamp: new Date().getTime()
      }
      set(ref(database,`/rooms/${path}/chats/${new Date().getTime()}-${user.username}`), payload)
      remove(ref(database, `/rooms/${path}/users/online/${user.uid}`))
      setChatRoomObj(null)
      setAlreadyLeft(true)
      setMainTab("home")
    })
  }

  // Adds room to myRooms
  function addToMyRooms() {
    fetch('/api/user').then((res) => res.json())
    .then((user) => {
      set(ref(database,`/users/${user.uid}/rooms/${chatRoomObj.name}-${chatRoomObj.timestamp}`), {
        name: chatRoomObj.name,
        path: chatRoomObj.path,
        timestamp: chatRoomObj.timestamp,
        description: chatRoomObj.description,
        longitude: chatRoomObj.longitude,
        latitude: chatRoomObj.latitude,
      })
    })
    setIsMyRoom(true)
  }

  // Deletes saved room from myRooms
  function removeFromMyRooms() {
    fetch('/api/user').then((res) => res.json())
    .then((user) => {
      remove(ref(database,`/users/${user.uid}/rooms/${chatRoomObj.name}-${chatRoomObj.timestamp}`))
    })
    setIsMyRoom(false)
  }

  // Fires to tell other uses that you are leaving the room
  useBeforeunload(() => {
    fetch('/api/user').then((res) => res.json())
      .then((user) => {
        if (chatRoomObj && mainTab == "chat") {
          var payload = {
            body: "left",
            user: user.username,
            isSystem: true,
            timestamp: new Date().getTime()
          }
          set(ref(database,`/rooms/${chatRoomObj.path+"/"+chatRoomObj.name+"-"+chatRoomObj.timestamp}/chats/${new Date().getTime()}-${user.username}`), payload)
          remove(ref(database, `/rooms/${chatRoomObj.path+"/"+chatRoomObj.name+"-"+chatRoomObj.timestamp}/users/online/${user.uid}`))
        }
      })
  });

  return (
    <div className="grid grid-cols-4 auto-cols-max overflow-hidden">
      {/* Left Side of Page */}
      <div className="col-span-3 h-dvh">
        {/* Header */}
        <div className="m-2 rounded-lg h-[63px] bg-white shadow-2xl grid grid-cols-2 p-1">
          <div className='h-[60px]'>
            <a href="/"><img src="logos/logo_transparent_inverse.png" className='h-[60px]'/></a>
          </div>
          <div className='h-[60px] p-4'>
            {(mainTab == "chat" && isMyRoom == false) && <a onClick={() => {addToMyRooms()}} className="p-2 cursor-pointer bg-[#dee0e0] bg-cyan-500 text-white font-bold rounded-full mr-5">Add to &quot;My Rooms&quot;</a>}
            {(mainTab == "chat" && isMyRoom == true) && <a onClick={() => {removeFromMyRooms()}} className="p-2 cursor-pointer bg-[#dee0e0] bg-cyan-500 text-white font-bold rounded-full mr-5">Remove from &quot;My Rooms&quot;</a>}
            {mainTab == "chat" && <a onClick={() => {closeChatRoom(chatRoomObj)}} className="p-2 cursor-pointer bg-[#dee0e0] bg-cyan-500 text-white font-bold rounded-full mr-5">Close Chat</a>}
            <a href="/api/signout" className="p-2 cursor-pointer bg-[#dee0e0] bg-cyan-500 text-white font-bold rounded-full">Sign Out</a>
          </div>
        </div>
        {/* Main Page Section */}
        <div className="mr-2 h-[calc(100%-110px)]">
          {(mainTab == "home" && !loadingLoc) && <MainTabHome loc={location}/>}
          {(mainTab == "home" && loadingLoc) && <MainTabHome loc={null}/>}
          {mainTab == "chat" && <MainTabChatRoom roomObj={chatRoomObj}/>}
        </div>
      </div>
      {/* Sidebar (Right Side of Page) */}
      {mainTab == "home" &&
      <div className="h-dvh">
          <div className="bg-white shadow-2xl rounded-lg m-2 h-[98%]">
              <div className='p-2'>
                <div className='p-1 rounded-lg grid grid-cols-3 bg-white'>
                  <div className={tab == "nearby"? 'select-none p-1 cursor-pointer rounded-lg hover:bg-[#C0C0C0] bg-[#D3D3D3]': 'select-none p-1 cursor-pointer rounded-lg hover:bg-[#C0C0C0]'} onClick={() => {setTab("nearby")}}>Nearby</div>
                  <div className={tab == "rooms"? 'select-none p-1 cursor-pointer rounded-lg hover:bg-[#C0C0C0] bg-[#D3D3D3]': 'select-none p-1 cursor-pointer rounded-lg hover:bg-[#C0C0C0]'} onClick={() => {setTab("rooms")}}>My Rooms</div>
                  <div className={tab == "create"? 'select-none p-1 cursor-pointer rounded-lg hover:bg-[#C0C0C0] bg-[#D3D3D3]': 'select-none p-1 cursor-pointer rounded-lg hover:bg-[#C0C0C0]'} onClick={() => {setTab("create")}}>Create</div>
                </div>
              </div>
              {(tab == "nearby") && <div className='overflow-y-auto h-[90%]'>
                <div>
                  {(!nearby && !loadingNearby) && <div>No Nearby Rooms<br/>Create One?</div>}
                  {loadingNearby && <div>Loading...</div>}
                  {nearby}
                </div>
              </div>}
              {tab == "rooms" && <div className='overflow-y-auto h-[90%]'>
                <div>
                  {isRoomLoading && <div>Loading</div>}
                  {(!myRooms && !isRoomLoading) && <div>No User Saved Rooms</div>}
                  {myRooms}
                </div>
              </div>}
              {(tab == "create" && !loadingLoc) && <CreateRoom loc={location}/>}
              {(tab == "create" && loadingLoc) && <div>Loading...</div>}
          </div>
      </div> }
      {(mainTab == "chat") && 
      <div className="h-dvh">
        <div className="m-2 h-[98%] grid grid-cols-1">
          <div className='bg-white rounded-lg m-2 shadow-2xl relative'>
              <div className='w-[100%] h-[100%] opacity-50 absolute rounded-lg z-10'>
                <Geo loc={{latitude: parseFloat(chatRoomObj.latitude.toFixed(2)), longitude: parseFloat(chatRoomObj.longitude.toFixed(2))}} zoom={12} movable={false} marker={false}/>
              </div>
              <div className='z-10 top-0 left-0 w-[100%] h-[100%] absolute text-left pl-3 pt-2'>
                <span className='font-bold text-[24px]'>{chatRoomObj.name}</span><br/>
                {chatRoomObj.description}
              </div>
          </div>
          <div className='bg-white rounded-lg m-2 shadow-2xl'>
            <div>
              Online Members
            </div>
            {chatroomOnline}
          </div>
          <div className='bg-white rounded-lg m-2 shadow-2xl'>
            <div>
              All Members
            </div>
            {chatroomUsers}
          </div>
        </div>
      </div>
      }
      {(mainTab == "profile") && 
      <div className="h-dvh">
        <div className=" bg-white m-2 h-[98%]">
          Profile
        </div>
      </div>
      }
    </div>
  )
}

export default Home;