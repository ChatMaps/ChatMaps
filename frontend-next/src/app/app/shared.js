export function Header() {
    return (
        <div className="m-2 rounded-lg h-[60px] bg-white shadow-2xl">
            <a href="/"><img src="/logos/logo_transparent_inverse.png" className="h-[60px]"/></a>
        </div>
    )
}

export function Sidebar() {
    return (
        <div className="h-dvh">
            <div className="bg-white shadow-2xl rounded-lg m-2 h-[98%]">
                Sidebar
            </div>
        </div>
      
    )
  }
