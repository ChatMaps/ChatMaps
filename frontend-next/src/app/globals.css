@tailwind base;
@tailwind components;
@tailwind utilities;

body {
  background-color: aliceblue;
  text-align: center;
  text-wrap: pretty;
}

button {
  background: #dee0e0;
  border-color: black;
  border: 5px;
  border-radius: 5px;
  padding: 5px;
  margin: 5px;
  filter: drop-shadow(0 25px 25px rgb(0 0 0 / 0.15));
}

button:hover {
  background-color: rgb(166, 166, 166);
}

input {
  border: 1px solid black;
  border-radius: 4px;
  padding: 10px 10px;
  margin: 5px;
}

input.err {
  border: 2px solid red;
  border-radius: 4px;
  padding: 10px 10px;
  margin: 5px;
}
