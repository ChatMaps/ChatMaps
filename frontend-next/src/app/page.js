"use client"
import { useState, useEffect } from 'react'

import { auth, database } from "./api/firebase-config";
import { ref, get} from "firebase/database";
import {onAuthStateChanged} from "firebase/auth"


function Home() {
    const [isLoadingLoc, setLoadingLoc] = useState(true)
    const [roomCount, setRoomCount] = useState(null)
    const [isAuthenticated, setAuth] = useState(false)
    const [userID, setUserID] = useState(null)

    // Authentication
  useEffect(() => {
    onAuthStateChanged(auth, (user) => {
      if (user) {
        setUserID(user.uid)
        setAuth(true)
      } else {
        setAuth(false)
      }
      })
  }, [])

    useEffect(() => {
        if('geolocation' in navigator) {
            // Retrieve latitude & longitude coordinates from `navigator.geolocation` Web API
            navigator.geolocation.getCurrentPosition(({ coords }) => {
              var path = String(coords.latitude.toFixed(2)).replace(".","")+"/"+String(coords.longitude.toFixed(2)).replace(".","")
              get(ref(database, `/rooms/${path}`)).then((snapshot) => {
                if (snapshot.exists()) {
                    var count = 0
                    for (var room in snapshot.val()) {
                        count += 1
                    }
                    setRoomCount(count)
                } else {
                    console.log("No rooms nearby")
                    setRoomCount(0)
                }
                setLoadingLoc(false)
              });
        });
          }
    })
    return (
        <div>
            <div className="grid h-screen place-items-center">
                <div>
                    <img src="logos/logo_transparent_inverse.png"/>
                    <span className="text-[36px]">
                        Chat with friends!
                    </span>
                    <div className="m-5">
                        {(!isAuthenticated) && 
                            <div>
                                <a href="/login"><button className="bg-cyan-500 text-white font-bold py-2 px-4 rounded-full">Login</button></a>
                                <a href="/register"><button className="bg-cyan-500 text-white font-bold py-2 px-4 rounded-full">Sign Up</button></a>
                            </div>
                        }
                        {isAuthenticated &&  <a href="/app"><button className="bg-cyan-500 text-white font-bold py-2 px-4 rounded-full">Continue to App</button></a>}
                        {(!isLoadingLoc && roomCount == 1) && <div className='text-[24px] pt-10'>Join others in {roomCount} room near you!</div>}
                        {(!isLoadingLoc && roomCount != 1 && roomCount != 0) && <div className='text-[24px] pt-10'>Join others in {roomCount} rooms near you!</div>}
                        {(isLoadingLoc || (roomCount == 0 && !isLoadingLoc)) && <div className='text-[24px] pt-10'>Start the conversation today!</div>}
                    </div>
                </div>
            </div>
        </div>
    )
}

export default Home;