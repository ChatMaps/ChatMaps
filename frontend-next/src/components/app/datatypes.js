import Link from "next/link"
import { useEffect, useState } from "react";

// Icons
import PersonIcon from '@mui/icons-material/Person';

// Colors for Messages
const userColors = [
  "#ff80ed",
  "#065535",
  "#133337",
  "#ffc0cb",
  "#e6e6fa",
  "#ffd700",
  "#ffa500",
  "#0000ff",
  "#1b85b8",
  "#5a5255",
  "#559e83",
  "#ae5a41",
  "#c3cb71",
];

let dateOptions = {
  weekday: "long",
  year: "numeric",
  month: "short",
  day: "numeric",
  hour: "2-digit",
  minute: "2-digit",
};

/**
 * Rich Message Formatting
 * @param {String} message - Message to Format
 * @returns {String} - Formatted Message (IN HTML)
 */
function RMF(message) {
  var URLREGEX = /[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)/g;
  var URLmatch = message.match(URLREGEX);
  if (URLmatch) {
    for (var i = 0; i < URLmatch.length; i++) {
      var link = (<span>
        {message.split(URLmatch[i])[0]}
        <Link href={"https://"+URLmatch[i]} target="_blank" className="hover:underline">{URLmatch[i]}</Link>
        {message.split(URLmatch[i])[1]}
      </span>)
      message = link
    }
  }
  return message
}
/**
 * Grabs Window Size
 * @returns {Object} - Window Size Object (width, height)
 */
export function useWindowSize() {
  const [windowSize, setWindowSize] = useState({
    width: undefined,
    height: undefined,
  });

  useEffect(() => {
    function handleResize() {
      setWindowSize({
        width: window.innerWidth,
        height: window.innerHeight,
      });
    }
    window.addEventListener("resize", handleResize);
    handleResize();
    return () => window.removeEventListener("resize", handleResize);
  }, []);
  return windowSize;
}

/**
 * Generates Color based on string hash
 * @param {String} user_str - Username / String for hashing
 * @returns {String} - Color Hex Code Index in userColors
 */
const generateColor = (user_str) => {
  // hashes username for consistent colors, maybe all functionality to pick color later
  let hash = 0;
  for (let i = 0; i < user_str.length; i++) {
    hash = user_str.charCodeAt(i) + (hash * 32 - hash);
  }
  const index = Math.abs(hash) % userColors.length;
  return index;
};


/**
 * Chat Message Object
 * @props {JSON} chatObj - Chat Object
 * @returns {Object} - Chat Message Component
 */
export function Chat({ chatObj }) {
  var message = RMF(chatObj.body)
  return (
    <div className="width-[100%] bg-white rounded-lg mt-1 text-left p-1 grid grid-cols-2 mr-2">
      <div>
        <span style={{ color: userColors[generateColor(chatObj.user)] }}>
          <Link href={`/user?uid=${chatObj.uid}`}
          className="hover:font-bold cursor-pointer">
            {chatObj.user}
          </Link>
        </span>
        : {message}
      </div>
      <div className="text-right text-[#d1d1d1]">
        {new Date(chatObj.timestamp).toLocaleString(dateOptions)}
      </div>
    </div>
  );
}


/**
 * System Chat Message Object
 * @prop {JSON} chatObj - Chat Object
 * @returns {Object} - System Chat Message Component
 */
export function SystemMessage({ chatObj }) {
  return (
    <div className="width-[100%] bg-white rounded-lg mt-1 text-left p-1 grid grid-cols-2 mr-2">
      <div className="text-[#d1d1d1]">
        <span style={{ color: userColors[generateColor(chatObj.user)] }}>
          <Link href={`/user?uid=${chatObj.uid}`}
          className="hover:font-bold cursor-pointer">
            {chatObj.user}
          </Link>
        </span>{" "}
        has {chatObj.body} the room.
      </div>
      <div className="text-right text-[#d1d1d1]">
        {new Date(chatObj.timestamp).toLocaleString(dateOptions)}
      </div>
    </div>
  );
}

/**
 * Member Object for Sidebar
 * @prop {JSON} memberObj - Member Object
 * @returns {Object} - Member Component
 */
export function Member({ memberObj }) {
  return (
    <Link href={"/user?uid=" + memberObj.uid}>
      <div className="cursor-pointer g-[aliceblue] rounded-lg m-3 shadow-xl p-2">
        {memberObj.username}
      </div>
    </Link>
  );
}

/**
 * Chat Room Object for Sidebar
 * @prop {JSON} roomObj - Room Object
 * @returns {Object} - Chat Room Component
 */
export function ChatRoomSidebar({ roomObj }) {
  if ("users" in roomObj) {
    if ("online" in roomObj.users) {
      var roomOnline = Object.keys(roomObj.users.online).length
    } else {
      var roomOnline = 0
    }
    if ("all" in roomObj.users) {
      var roomTotal = Object.keys(roomObj.users.all).length
    } else {
      var roomTotal = 0
    }
  } else {
    var roomOnline = 0
    var roomTotal = 0
  }
  return (
    <div className="border-[black] border-1 shadow-lg p-2 m-2 rounded-lg cursor-pointer">
      <Link href={`/chat?room=${roomObj.path}/${roomObj.name}-${roomObj.timestamp}`}>
        <div className="grid grid-cols-3">
          <div>
            <PersonIcon/>
            <div>{roomOnline} / {roomTotal}</div>
          </div>
          <div className="col-span-2">
            <div className="font-bold">{roomObj.name}</div>
            <div className="italic">{roomObj.description}</div>
          </div>
        </div>
      </Link>
    </div>
  );
}

// This will be removed once dateOptions is no longer used in this file
export { dateOptions };