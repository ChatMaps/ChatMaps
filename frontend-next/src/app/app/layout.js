import { Inter } from "next/font/google";
import "../globals.css";
import { Header, Sidebar } from "./shared"

const inter = Inter({ subsets: ["latin"] });

export const metadata = {
  title: "ChatMaps: Home",
  description: "ChatMaps: Social Media for College Students",
};

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <div className="grid grid-cols-4 auto-cols-max overflow-hidden">
          <div className="col-span-3 h-page">
            <Header/>
            {children}
          </div>
          <Sidebar/>
        </div>
        </body>
    </html>
  );
}
