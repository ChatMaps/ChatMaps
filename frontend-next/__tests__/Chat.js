import { Chat } from "../src/components/app/datatypes";
import renderer from 'react-test-renderer';

var exampleChatObj = {
    body: "Hello, World!",
    isSystem: false,
    timestamp: 1710527946340,
    user: "TestUser"
}

var exampleUser = {
    uid: "123456",
    username: "TestUser",
    lastOnline: true
}

it('Chat Renders Correctly', () => {
    const tree = renderer
        .create((<Chat chatObj={exampleChatObj} user={exampleUser} />))
        .toJSON();
    expect(tree).toMatchSnapshot();
});