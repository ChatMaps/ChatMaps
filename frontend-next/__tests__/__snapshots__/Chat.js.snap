// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Renders Correctly 1`] = `
<div
  className="width-[100%] bg-white rounded-lg mt-1 text-left p-1 grid grid-cols-2 mr-2"
>
  <div>
    <span
      className="mr-[5px]"
      style={
        {
          "color": "#133337",
        }
      }
    >
      <a
        className="hover:font-bold cursor-pointer"
        href="/user?uid=undefined"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        TestUser
      </a>
    </span>
    Hello, World!
  </div>
  <div
    className="text-right text-[#d1d1d1]"
  >
    3/15/2024, 2:39:06 PM
  </div>
</div>
`;
